pennylane/qnode_new/jacobian.py

```diff
@@ -0,0 +1,528 @@
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import ObservableReturnTypes, CV
+from pennylane.utils import _flatten, unflatten, _inv_dict
+from pennylane.qnode_new.qnode import QNode, QuantumFunctionError
+
+
+
+class JacobianQNode(QNode):
+    """Quantum node that can be differentiated with respect to its positional parameters.
+    """
+    interface = None  #: str, None: automatic differentiation interface used by the node, if any
+
+    def __init__(self, func, device, mutable=True, properties=None):
+        super().__init__(func, device, mutable=mutable, properties=properties)
+
+        #: dict[int, str]: map from free parameter index to the gradient method to be used with that parameter
+        self.par_to_grad_method = None
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<JNode: device='{}', func={}, wires={}, interface={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires, self.interface)
+
+    def _construct(self, args, kwargs):
+        """Constructs the quantum circuit graph by calling the quantum function.
+
+        Like :meth:`.QNode._construct`, additionally determines the best gradient computation method
+        for each free parameter.
+        """
+        super()._construct(args, kwargs)
+        self.par_to_grad_method = {k: self._best_method(k) for k in self.variable_deps}
+
+        temp = [str(ob) for ob in self.circuit.observables if ob.return_type is ObservableReturnTypes.Sample]
+        if temp:
+            raise QuantumFunctionError("Circuits that include sampling can not be differentiated. "
+                                       "The following observables include sampling: {}".format('; '.join(temp)))
+
+    def _best_method(self, idx):
+        """Determine the correct partial derivative computation method for a free parameter.
+
+        Use the analytic method iff every gate that depends on the parameter supports it.
+        If not, use the finite difference method only (one would have to use it anyway).
+
+        Note that if even one dependent Operation does not support differentiation,
+        we cannot differentiate with respect to this parameter at all.
+
+        Args:
+            idx (int): free parameter index
+        Returns:
+            str: partial derivative method to be used
+        """
+        # TODO make sure Identity is recognized as gaussian
+
+        # pylint: disable=too-many-return-statements
+        def best_for_op(op):
+            "Returns the best gradient method for the Operation op."
+            # for discrete operations, other ops do not affect the choice
+
+            if not isinstance(op, CV):
+                return op.grad_method
+
+            # for CV ops it is more complicated
+            if op.grad_method == "A":
+                # op is Gaussian and has the heisenberg_* methods
+
+                obs_successors = self._op_descendants(op, 'E')
```

Comment:
Are we now using the circuitgraph to determine `_op_descendents`? If not, we should try to do so